Add new meshGateway key for Connect Mesh Gateways

lkysow opened the pull request:
- [x] consul-k8s changes merged
- [x] test with bootstrap acl
        - NOTE: this worked *but* the gateways fail unless there's a connect-enabled service running **and** for some reason the connect inject container kept getting `Error logging in: Unexpected response code: 403 (rpc error making call: rpc error making call: Permission denied)` errors...
- [x] figure out how best to specify wan-address
- [x] finish Service config
- [x] decide on default resources
- [x] liveness/readiness probes (should be able to use localhost tcp 443)
- [x] docs for all the options in values.yaml
- [x] deregistration on exit
- [x] add pod security policy
- [x] website docs
- [x] test with multi-dc
- [x] tests

lkysow commented on templates/client-daemonset.yaml:
```diff
@@ -101,7 +101,7 @@ spec:
                 -bind=0.0.0.0 \
                 -client=0.0.0.0 \
-                {{- if .Values.client.grpc }}
+                {{- if or .Values.client.grpc .Values.meshGateway.enabled }}
```
gateways require gRPC. I can either do this or require `client.grpc` to be set to true. Requiring them to set this in client.grpc is cleaner because then they don't get grpc unexpectedly. But then we should find a way to require that if meshGateway.enabled is true, that client.grpc is true.

lkysow replied:
Okay, I'll do that in a separate PR.

lkysow replied:
I'm actually going to remove this because it feels wrong to set grpc to true automatically. Instead I'll document that it must be set to true for gateways to work *and* I've found a neat trick to have helm fail if these things aren't set properly. I'll tag you on it when it's committed.

adilyse commented on values.yaml:
```diff
@@ -167,7 +167,7 @@ client:
 
   # grpc should be set to true if the gRPC listener should be enabled.
-  # This should be set to true if connectInject is enabled.
+  # This will be set to true if connectInject or meshGateway is enabled.
```
```suggestion
  # This will automatically be set to true if connectInject or meshGateway is enabled.
```

This statement isn't entirely true unless we fix it so that connectInject turns this on as well (mentioned in the comment on the client daemonset file).

lkysow replied:
✅

adilyse commented on templates/mesh-gateway-deployment.yaml:
```diff
@@ -0,0 +1,162 @@
+{{- if and .Values.meshGateway.enabled }}
```
```suggestion
{{- if .Values.meshGateway.enabled }}
```

adilyse commented on templates/mesh-gateway-clusterrolebinding.yaml:
```diff
@@ -0,0 +1,20 @@
+{{- if and .Values.meshGateway.enabled }}
```
```suggestion
{{- if .Values.meshGateway.enabled }}
```

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+    port: 443
+
+    # Optional YAML string for additional annotations.
+    annotations: |
```
This and `additionalSpec` should be null, though it would be good to have an example of formatting in the comment above it. Elsewhere in the Helm chart we've set the standard that none of the options are commented out, but the ones not in use are `null` or empty. It would be good to be consistent.

lkysow replied:
✅

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+    additionalSpec: |
+
+  # Envoy image to use.
+  imageEnvoy: envoyproxy/envoy:v1.8.0
```
Per other discussion, we should be using envoy 1.10, or at least 1.9.1+

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+
+    # If true, each Gateway Pod will advertise its NodeName
+    # (as provided by the Kubernetes downward API) as the wan address.
+    # Again, this is useful if the node names are DNS entries that are
```
```suggestion
    # This is useful if the node names are DNS entries that are
```

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+    # (as provided by the Kubernetes downward API) as the wan address.
+    # Again, this is useful if the node names are DNS entries that are
+    # routable from other DCs.
+    # The port key will be used as the port for the wan address.
```
Same comment about `key`.

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+    # Users must ensure that this address routes to the Gateway pods,
+    # for example via a DNS
+    # entry that routes to the Service fronting the Deployment.
+    # The port key will be used as the port for the wan address.
```
Same comment about `key`.

lkysow replied:
✅

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+
+    # If set, this will be the host of the wan address for all gateways.
+    # Users must ensure that this address routes to the Gateway pods,
+    # for example via a DNS
```
Some odd line wrapping in this comment.

lkysow replied:
✅

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+    # Must set useNodeIP to false.
+    # useNodeName: false
+
+    # If set, this will be the host of the wan address for all gateways.
```
I'm having a hard time parsing "will be the host of the wan address for all gateways." Is there another way to phrase this to make it clearer?

lkysow replied:
Changed to: "# If set, each gateway Pod will use this host as its wan address."

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+    # If true, each Gateway Pod will advertise its NodeIP
+    # (as provided by the Kubernetes downward API) as the wan address.
+    # This is useful if the node IPs are routable from other DCs.
+    # The port key will be used as the port for the wan address.
```
What do you mean by `port key`? Something like this? 

```suggestion
    # The port configured above will be used as the port for the wan address.
```

We use `key` elsewhere in this file in several different ways, so overloading it here as well will get confusing.

lkysow replied:
✅

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+    # Must set useNodeIP and useNodeName to false.
+    # host: myhostname.com
+
+  # The service key configures the Service that fronts the Gateway Deployment.
```
Same comment about `key`.

lkysow replied:
✅

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+  dnsPolicy: ClusterFirst
+
+  # Override the default 'mesh-gateway' service name registered in Consul.
+  # Cannot be used if bootstrapACLs is true since the ACL token generated
```
There should definitely be a test for this, since it would be good to have a hard requirement on this in the templating.

lkysow replied:
How can I do a hard requirement?

adilyse replied:
Where the `consulServiceName` is used and `bootstrapACLs` is true, you can require that the service name is equal to `mesh-gateway` or that section is not rendered.

lkysow replied:
I'm going to do this via a `fail` call. I'll tag you on it, lmk what you think.

adilyse commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+  # Optional hostPort for the gateway to be exposed on.
+  # This can be used with wanAddress.port and wanAddress.useNodeIP
+  # to expose the gateways directly from the node.
+  # hostPort: null
```
Does this potentially conflict with the `hostNetwork` option?

lkysow replied:
A bit, they have to set it to the containerPort if using `hostNetwork: true`. I'll add that to the docs.

Kube docs:
> If HostNetwork is specified, this must match ContainerPort.

adilyse commented on templates/mesh-gateway-deployment.yaml:
```diff
@@ -0,0 +1,162 @@
+      - name: consul-bin
+        emptyDir: {}
+      hostNetwork: {{ .Values.meshGateway.hostNetwork }}
+      dnsPolicy: {{ .Values.meshGateway.dnsPolicy }}
```
Should this be option in case someone sets the `dnsPolicy` to null?

lkysow replied:
✅

adilyse commented on templates/mesh-gateway-deployment.yaml:
```diff
@@ -0,0 +1,162 @@
+          valueFrom:
+            fieldRef:
+              fieldPath: status.podIP
+        - name: NODE_NAME
```
This should be optional, only if `.Values.meshGateway.wanAddress.useNodeName` is true.

lkysow replied:
Is there a downside to having it here regardless? My thought was that it's less conditionals == less complexity.

lkysow commented:
> Big open question from me: Does the mesh gateway need to have Connect and/or central config enabled to work correctly? If so, there are some places that need slightly different logic.
Yes both connect and central config are needed. Have you commented on all the places that need slightly different logic or does that warrant further discussion?


> Also: Do the gateways need to run on a node with a client available? If so, it would be good to note that somewhere in the affinity documentation so that folks don't accidentally partition things in an unusable way.
Yes they do, good point.

lkysow commented on values.yaml:
```diff
@@ -387,3 +387,116 @@ connectInject:
+  # be proxied via the local gateway. Use 'local' if services cannot directly
+  # connect to the other datacenter's gateway and so must proxy through the local
+  # gateway, i.e. if only the gateways can communicate between datacenters.
+  mode: remote
```
✅

lkysow commented on templates/mesh-gateway-clusterrole.yaml:
```diff
@@ -0,0 +1,24 @@
+  verbs:
+    - get
+{{- else }}
+rules: []
```
✅

adilyse commented on templates/mesh-gateway-deployment.yaml:
```diff
@@ -0,0 +1,169 @@
+              valueFrom:
+                fieldRef:
+                  fieldPath: status.podIP
+            - name: NODE_NAME
```
We shouldn't add unnecessary things to a deployment if it's not needed, even if it adds a bit of complexity. Tests are a good way to make sure the logic does the right thing.

lkysow replied:
✅

lkysow replied:
yep, I thought I added that in there but must have got deleted somehow

adilyse commented on test/unit/mesh-gateway-deployment.bats:
```diff
@@ -0,0 +1,220 @@
+  [ "${actual}" = "true" ]
+}
+
+@test "meshGateway/Deployment: annotations can be set" {
```
Might be good to make sure that setting more than one annotation will be added and formatted correctly.

lkysow replied:
We can't set multiline strings via `--set` unfortunately. I could write a `values.yaml` file to disk I guess? Might be overkill, what do you think?

adilyse commented on values.yaml:
```diff
@@ -387,3 +389,130 @@ connectInject:
+meshGateway:
+  # If mesh gateways are enabled, a Deployment will be created that runs
+  # gateways and Consul Connect will be configured to use gateways.
+  # See {todo doc link} for more information.
```
Reminder that this is a todo link.

lkysow replied:
✅

adilyse commented on values.yaml:
```diff
@@ -387,3 +389,130 @@ connectInject:
+    # useNodeIP and useNodeName must be false.
+    host: ""
+
+  # The service key configures the Service that fronts the Gateway Deployment.
```
```suggestion
  # The service option configures the Service that fronts the Gateway Deployment.
```

lkysow replied:
✅

lkysow commented on templates/mesh-gateway-deployment.yaml:
```diff
@@ -0,0 +1,172 @@
+{{- if .Values.meshGateway.enabled }}
+{{- if not .Values.connectInject.enabled }}{{ fail "connectInject.enabled must be true" }}{{ end -}}
```
@adilyse this is what I'm doing to enforce requirements. What do you think?

lkysow replied:
Yeah it's not great but not sure what else to do. I think this is better than not rendering the template since they'll get an error and can fix it rather than wondering why there's no gateways spinning up

adilyse replied:
Agreed!

lkysow commented on templates/mesh-gateway-deployment.yaml:
```diff
@@ -0,0 +1,172 @@
+                -wan-address="${HOST_IP}:{{ .Values.meshGateway.wanAddress.port }}" \
+                {{- end }}
+                {{- if .Values.meshGateway.consulServiceName }}
+                {{- if .Values.global.bootstrapACLs }}{{ fail "if global.bootstrapACLs is true, meshGateway.consulServiceName cannot be set" }}{{ end }}
```
also here

adilyse commented on test/unit/mesh-gateway-deployment.bats:
```diff
@@ -0,0 +1,505 @@
+  [[ "${actual}" =~ '-wan-address="myhost:4444"' ]]
+}
+
+@test "meshGateway/Deployment: fails if consulServiceName is set and bootstrapACLs is true" {
```
As mentioned elsewhere, it would be nice for this to work if someone happened to set the service name to `mesh-gateway`. A test for that case would be good.

lkysow replied:
✅

adilyse commented on test/unit/mesh-gateway-clusterrolebinding.bats:
```diff
@@ -0,0 +1,37 @@
+  [ "${actual}" = "true" ]
+}
+
+@test "meshGateway/ClusterRoleBinding: subject name is correct" {
```
This test fails locally for me:

```
`[ "${actual}" = "release-name-consul-mesh-gateway" ]' failed

RELEASE-NAME-consul-mesh-gateway
```

lkysow replied:
Maybe a helm version difference?

```
Client: &version.Version{SemVer:"v2.14.1", GitCommit:"5270352a09c7e8b6e8c9593002a73535276507c0", GitTreeState:"clean"}
	Server: &version.Version{SemVer:"v2.14.1", GitCommit:"5270352a09c7e8b6e8c9593002a73535276507c0", GitTreeState:"clean"}
```

lkysow replied:
I've set it explicitly via `--name`

adilyse commented on test/unit/mesh-gateway-podsecuritypolicy.bats:
```diff
@@ -0,0 +1,25 @@
+      . | tee /dev/stderr |
+      yq 'length > 0' | tee /dev/stderr)
+  [ "${actual}" = "true" ]
+}
```
Again, it would be good to make sure this is disabled if connectInject and client.grpc are not enabled.

lkysow replied:
The whole chart will fail in that case

adilyse commented on test/unit/mesh-gateway-podsecuritypolicy.bats:
```diff
@@ -0,0 +1,25 @@
+  [ "${actual}" = "false" ]
+}
+
+@test "meshGateway/PodSecurityPolicy: enabled with meshGateway.enabled and global.enablePodSecurityPolicies=true" {
```
```suggestion
@test "meshGateway/PodSecurityPolicy: enabled with meshGateway, connectInject and client.grpc enabled and global.enablePodSecurityPolicies=true" {
```

adilyse commented on test/unit/mesh-gateway-service.bats:
```diff
@@ -0,0 +1,186 @@
+  [ "${actual}" = "false" ]
+}
+
+@test "meshGateway/Service: disabled with meshGateway.enabled=true" {
```
```suggestion
@test "meshGateway/Service: disabled by default with meshGateway, connectInject and client.grpc enabled" {
```

adilyse commented on test/unit/mesh-gateway-serviceaccount.bats:
```diff
@@ -0,0 +1,25 @@
+  [ "${actual}" = "false" ]
+}
+
+@test "meshGateway/ServiceAccount: enabled with meshGateway.enabled = true" {
```
```suggestion
@test "meshGateway/ServiceAccount: enabled with meshGateway, connectInject and client.grpc enabled" {
```

adilyse commented on test/unit/mesh-gateway-serviceaccount.bats:
```diff
@@ -0,0 +1,25 @@
+      yq 'length > 0' | tee /dev/stderr)
+  [ "${actual}" = "true" ]
+}
+
```
Should test that this is disabled with connectInject and/or client.grpc turned off.

lkysow replied:
The whole chart will fail in that case

adilyse commented on test/unit/server-configmap.bats:
```diff
@@ -111,3 +111,48 @@ load _helpers
+      --set 'connectInject.centralConfig.enabled=true' \
+      --set 'connectInject.centralConfig.proxyDefaults="{\"hello\": \"world\"}"' \
+      --set 'meshGateway.enabled=true' \
+      --set 'meshGateway.globalMode=' \
```
Should this be set to `null`? Or `local`?

lkysow replied:
If set to null or empty string then it shouldn't be set. Added a test for null.

lkysow commented on test/unit/mesh-gateway-clusterrole.bats:
```diff
@@ -0,0 +1,77 @@
+      yq 'length > 0' | tee /dev/stderr)
+  [ "${actual}" = "true" ]
+}
+
```
The whole chart will fail to render if that's not the case. That's tested in the deployment bats tests.

lkysow commented on values.yaml:
```diff
@@ -408,3 +409,134 @@ connectInject:
+  # to expose the gateways directly from the node.
+  # If hostNetwork is true, this must be null or set to the same port as
+  # containerPort.
+  hostPort: null
```
✅

lkysow commented on values.yaml:
```diff
@@ -408,3 +409,134 @@ connectInject:
+  # Can be set to either "remote", "local", "none" or "".
+  # See https://consul.io/docs/connect/mesh_gateway.html#modes-of-operation for
+  # a description of each mode.
+  # If set, connectInject.centralConfig.enabled should be set to true so that
```
Made clearer.

lkysow commented on values.yaml:
```diff
@@ -408,3 +409,134 @@ connectInject:
+  # If mesh gateways are enabled, a Deployment will be created that runs
+  # gateways and Consul Connect will be configured to use gateways.
+  # See https://www.consul.io/docs/connect/mesh_gateway.html
+  # Requirements: consul >= 1.6.0 and consul-k8s >= 0.8.2 if using global.bootstrapACLs.
```
✅

lkysow commented on test/unit/mesh-gateway-clusterrolebinding.bats:
```diff
@@ -0,0 +1,37 @@
+      yq 'length > 0' | tee /dev/stderr)
+  [ "${actual}" = "true" ]
+}
+
```
the whole chart will fail to render in this case, it's tested in the deployment

lkysow commented on test/unit/mesh-gateway-clusterrole.bats:
```diff
@@ -0,0 +1,77 @@
+      yq -r '.rules[0].resources[0]' | tee /dev/stderr)
+  [ "${actual}" = "podsecuritypolicies" ]
+}
+
```
✅

lkysow commented on test/unit/mesh-gateway-deployment.bats:
```diff
@@ -0,0 +1,505 @@
+      --set 'meshGateway.enabled=true' \
+      --set 'connectInject.enabled=true' \
+      --set 'client.grpc=false' \
+      --set 'connectInject.enabled=true' \
```
✅

lkysow commented on test/unit/mesh-gateway-deployment.bats:
```diff
@@ -0,0 +1,505 @@
+  [ "${actual}" = "value" ]
+}
+
+@test "meshGateway/Deployment: hostNetwork can be set" {
```
✅

lkysow commented:
Updates:
- added `fail` if clients are disabled, tested
- moved annotations to Pod from Deployment, tested multiple annotations
- added conditional render for NODE_NAME
- added special case for setting `consulServiceName` to `mesh-gateway`, tested
- added headings for test cases
- added test for `hostNetwork` not set by default
- added test for `globalMode=null`


> Realized that there's yet another dependency to keep track off-- the logic also needs to enforce that clients are enabled for mesh gateways. This one is slightly more complicated because both true and - are valid values for client.enabled.

Added a check to the top of the deployment spec and tests.

> Is having only the one error on the deployment enough? Or should the logic checking for connectInject and client.grpc be attached to all mesh-gateway components?

One is enough because even if you're rendering a single template via `helm template` it will fail. Helm must be rendering everything each time:
```
$ helm template \
  -x templates/mesh-gateway-clusterrole.yaml \
  --set 'meshGateway.enabled=true' .
Error: render error in "consul/templates/mesh-gateway-deployment.yaml": template: consul/templates/mesh-gateway-deployment.yaml:2:46: executing "consul/templates/mesh-gateway-deployment.yaml" at <fail "connectInject.enabled must be true">: error calling fail: connectInject.enabled must be true
```

> It would be good to break up the test files with headers for the various components being tested as defined here.

Done
